### PR TITLE
New version of VOMS slave script

### PR DESCRIPTION
**Pull request by sustr4**
- reimplemented in Perl
- supports VOMS Roles and Group membership
- requires new (XML) version of input data

This is a squashed version of commits already present in Pull Request #55. That one can be closed and this one merged instead.

### CODE REVIEW

File: slave-new/process-voms/bin/process-voms.sh [hunk 1]
```diff
@@ -1,131 +1,12 @@
...
-
-	### Error codes ###
-	E_VOMS_DELETE=(50 'Error during VOMS delete-user operation')
-	E_VOMS_CREATE=(50 'Error during VOMS create-user operation')
-
-	FROM_PERUN="${WORK_DIR}/voms"
-
-	RETVAL=0
-
-	create_lock
```
Comment (stavamichal) [inline slave-new/process-voms/bin/process-voms.sh]: Code with creating lock should stay there. It is protection before Race condition between two and more slave scripts.

File: slave-new/process-voms/lib/process-voms.pl [hunk 2]
```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/perl
+
+# initialize parser and read the file
+use Sys::Syslog;
+use XML::Simple;
+use Text::CSV;
+use Data::Dumper;
+use Array::Utils qw(:all);
+use JSON::XS;
```
Comment (stavamichal) [inline slave-new/process-voms/lib/process-voms.pl]: Please add all not standard libraries to dependencies (just for slave-new).

File: slave-new/process-voms/lib/process-voms.pl [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+				effectCall "voms-admin --nousercert --vo $name assign-role \"$group\" \"$role\" \"$user->{'DN'}\" \"$user->{'CA'}\"",
+				"assigning Role \"$role\" to user \"$user->{'DN'}\" for Group \"$group\" in VO \"$name\"";
+			}
+		}
+	}
+}
+
+closelog();
+
+$retval;
```
Comment (stavamichal) [inline slave-new/process-voms/lib/process-voms.pl]: Is this correct? Variable retval is created on the line 98 but then nothing set it with any value, just return at the end of the script.

File: slave-new/process-voms/bin/process-voms.sh [hunk 4]
```diff
@@ -1,131 +1,12 @@
...
-			2)
-				log_msg E_VO_AC_FAIL;;
-			3)
-				log_msg E_VO_FAIL;;
-		esac
-	fi
+	EXECSCRIPT="${LIB_DIR}/${SERVICE}/process-voms.pl"
 
-	exit $RETVAL
+	$EXECSCRIPT -
```
Comment (stavamichal) [inline slave-new/process-voms/bin/process-voms.sh]: Why are you using "-"? And maybe use something like "exit $?" at the end to get information about return code of your perl script.

File: slave-new/process-voms/lib/process-voms.pl [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+### knownCA indicates whether a given CA is known to the VOMS server. It accepts the user structure
+#	%user		The user whose CA should be checked (DN, CA, email)
+#	%list		Reference to the list of known CAs
+sub knownCA {
+	$ca = shift;
+	$list = shift;
+
+        if( "$ca" ~~ @{$list}) {
```
Comment (stavamichal) [inline slave-new/process-voms/lib/process-voms.pl]: Be careful about smartmatch, it is deprecated and experimental operator in higher version of Perl so maybe something like:
no if $] >= 5.018, warnings => "experimental";
will be needed.

File: slave-new/process-voms/bin/process-voms.sh (hunk 4 above)
Comment (sustr4) [inline slave-new/process-voms/bin/process-voms.sh]: What do you suggest instead? I tried several options I though should work.

Comment (stavamichal) [inline slave-new/process-voms/bin/process-voms.sh]: For example: ./$EXECSCRIPT or "perl $EXECSCRIPT"

**Comment by sustr4** [comment]
All tests passed. This can be merged and carefully deployed.

**Pull request merged**